*** RemoteManager_server.cpp	2012-02-03 22:32:54.973556002 +0200
--- RemoteManager_server.cpp.patched	2012-02-03 22:39:02.509883282 +0200
***************
*** 220,233 ****
  	_return =  mgr->GetNodeType((::uint32 const) _homeId, (::uint8 const) _nodeId);
  	g_criticalSection.unlock();
    }
  
    void GetNodeNeighbors(UInt32_ListByte& _return, const int32_t _homeId, const int8_t _nodeId) {
  	Manager* mgr = Manager::Get();
  	g_criticalSection.lock();
! 	_return.retval =  mgr->GetNodeNeighbors((::uint32 const) _homeId, (::uint8 const) _nodeId, (::uint8**) &_return._nodeNeighbors);
  	g_criticalSection.unlock();
    }
  
    void GetNodeManufacturerName(std::string& _return, const int32_t _homeId, const int8_t _nodeId) {
  	Manager* mgr = Manager::Get();
  	g_criticalSection.lock();
--- 220,238 ----
  	_return =  mgr->GetNodeType((::uint32 const) _homeId, (::uint8 const) _nodeId);
  	g_criticalSection.unlock();
    }
  
    void GetNodeNeighbors(UInt32_ListByte& _return, const int32_t _homeId, const int8_t _nodeId) {
+     uint8* arr;
  	Manager* mgr = Manager::Get();
  	g_criticalSection.lock();
! 	_return.retval =  mgr->GetNodeNeighbors((::uint32 const) _homeId, (::uint8 const) _nodeId, (::uint8**) &arr);
  	g_criticalSection.unlock();
+     if (_return.retval > 0) {
+         for (int i=0; i<_return.retval; i++) _return._nodeNeighbors.push_back(arr[i]);
+         delete arr;
+     }
    }
  
    void GetNodeManufacturerName(std::string& _return, const int32_t _homeId, const int8_t _nodeId) {
  	Manager* mgr = Manager::Get();
  	g_criticalSection.lock();
***************
*** 646,659 ****
  	g_criticalSection.unlock();
  	return(function_result);
    }
  
    void GetAssociations(GetAssociationsReturnStruct& _return, const int32_t _homeId, const int8_t _nodeId, const int8_t _groupIdx) {
  	Manager* mgr = Manager::Get();
  	g_criticalSection.lock();
! 	_return.retval =  mgr->GetAssociations((::uint32 const) _homeId, (::uint8 const) _nodeId, (::uint8 const) _groupIdx, (::uint8**) &_return.o_associations);
  	g_criticalSection.unlock();
    }
  
    int8_t GetMaxAssociations(const int32_t _homeId, const int8_t _nodeId, const int8_t _groupIdx) {
  	Manager* mgr = Manager::Get();
  	g_criticalSection.lock();
--- 651,669 ----
  	g_criticalSection.unlock();
  	return(function_result);
    }
  
    void GetAssociations(GetAssociationsReturnStruct& _return, const int32_t _homeId, const int8_t _nodeId, const int8_t _groupIdx) {
+     uint8* o_associations;
  	Manager* mgr = Manager::Get();
  	g_criticalSection.lock();
! 	_return.retval =  mgr->GetAssociations((::uint32 const) _homeId, (::uint8 const) _nodeId, (::uint8 const) _groupIdx, (::uint8**) &o_associations);
  	g_criticalSection.unlock();
+     if (_return.retval > 0) {
+         for (int i=0; i<_return.retval; i++) _return.o_associations.push_back(o_associations[i]);
+         delete o_associations;
+     }
    }
  
    int8_t GetMaxAssociations(const int32_t _homeId, const int8_t _nodeId, const int8_t _groupIdx) {
  	Manager* mgr = Manager::Get();
  	g_criticalSection.lock();
***************
*** 720,733 ****
  	g_criticalSection.unlock();
  	return(function_result);
    }
  
    void GetAllScenes(GetAllScenesReturnStruct& _return) {
  	Manager* mgr = Manager::Get();
  	g_criticalSection.lock();
! 	_return.retval =  mgr->GetAllScenes((::uint8**) &_return._sceneIds);
  	g_criticalSection.unlock();
    }
  
    int8_t CreateScene() {
  	Manager* mgr = Manager::Get();
  	g_criticalSection.lock();
--- 730,748 ----
  	g_criticalSection.unlock();
  	return(function_result);
    }
  
    void GetAllScenes(GetAllScenesReturnStruct& _return) {
+     uint8* _sceneIds;
  	Manager* mgr = Manager::Get();
  	g_criticalSection.lock();
! 	_return.retval =  mgr->GetAllScenes((::uint8**) &_sceneIds);
  	g_criticalSection.unlock();
+     if (_return.retval>0) {
+         for (int i=0; i<_return.retval; i++) _return._sceneIds.push_back(_sceneIds[i]);
+         delete(_sceneIds);
+     }
    }
  
    int8_t CreateScene() {
  	Manager* mgr = Manager::Get();
  	g_criticalSection.lock();
***************
*** 815,828 ****
  	g_criticalSection.unlock();
  	return(function_result);
    }
  
    void SceneGetValues(SceneGetValuesReturnStruct& _return, const int8_t _sceneId) {
  	Manager* mgr = Manager::Get();
  	g_criticalSection.lock();
! 	_return.retval =  mgr->SceneGetValues((::uint8 const) _sceneId, _return.o_value.toValueID());
  	g_criticalSection.unlock();
    }
  
    void SceneGetValueAsBool(Bool_Bool& _return, const int8_t _sceneId, const RemoteValueID& _valueId) {
  	Manager* mgr = Manager::Get();
  	g_criticalSection.lock();
--- 830,845 ----
  	g_criticalSection.unlock();
  	return(function_result);
    }
  
    void SceneGetValues(SceneGetValuesReturnStruct& _return, const int8_t _sceneId) {
+     std::vector<OpenZWave::ValueID>  o_values;
  	Manager* mgr = Manager::Get();
  	g_criticalSection.lock();
! 	_return.retval =  mgr->SceneGetValues((::uint8 const) _sceneId, &o_values);
  	g_criticalSection.unlock();
+     for (int i=0; i< _return.retval; i++) _return.o_value.push_back(RemoteValueID(o_values[i]));
    }
  
    void SceneGetValueAsBool(Bool_Bool& _return, const int8_t _sceneId, const RemoteValueID& _valueId) {
  	Manager* mgr = Manager::Get();
  	g_criticalSection.lock();
***************
*** 972,983 ****
  	g_criticalSection.unlock();
  	return(function_result);
    }
  
    void SendAllValues() {
!     // Your implementation goes here
!     printf("SendAllValues\n");
    }
  
  };
  
  // int main(int argc, char **argv) {
--- 989,999 ----
  	g_criticalSection.unlock();
  	return(function_result);
    }
  
    void SendAllValues() {
!     send_all_values();
    }
  
  };
  
  // int main(int argc, char **argv) {
