*** gen-cpp/RemoteManager_server.cpp.orig	2012-02-26 23:58:12.427952190 +0200
--- gen-cpp/RemoteManager_server.cpp.patched	2012-02-26 23:48:03.155177508 +0200
***************
*** 15,25 ****
  using boost::shared_ptr;
  
  using namespace  ::OpenZWave;
  
  void BeginControllerCommand_callback(OpenZWave::Driver::ControllerState  arg1, void*  arg2) {
! 	// FIXME: fill in the blanks (sorry!)
  }
  
  class RemoteManagerHandler : virtual public RemoteManagerIf {
   public:
    RemoteManagerHandler() {
--- 15,32 ----
  using boost::shared_ptr;
  
  using namespace  ::OpenZWave;
  
  void BeginControllerCommand_callback(OpenZWave::Driver::ControllerState  arg1, void*  arg2) {
!     g_criticalSection.lock();
!     //
!     STOMP::hdrmap headers;
!     headers["ControllerState"] = to_string<uint16_t>(arg1, std::hex);
!     string empty = ""  ;
!     stomp_client->send(*notifications_topic, headers, empty);
!     //
!     g_criticalSection.unlock();
  }
  
  class RemoteManagerHandler : virtual public RemoteManagerIf {
   public:
    RemoteManagerHandler() {
***************
*** 869,882 ****
  	g_criticalSection.unlock();
  	return(function_result);
    }
  
    void SceneGetValues(SceneGetValuesReturnStruct& _return, const int8_t _sceneId) {
  	Manager* mgr = Manager::Get();
  	g_criticalSection.lock();
! 	_return.retval =  mgr->SceneGetValues((::uint8 const) _sceneId, _return.o_value.toValueID());
  	g_criticalSection.unlock();
    }
  
    void SceneGetValueAsBool(Bool_Bool& _return, const int8_t _sceneId, const RemoteValueID& _valueId) {
  	Manager* mgr = Manager::Get();
  	g_criticalSection.lock();
--- 876,891 ----
  	g_criticalSection.unlock();
  	return(function_result);
    }
  
    void SceneGetValues(SceneGetValuesReturnStruct& _return, const int8_t _sceneId) {
+     std::vector<OpenZWave::ValueID>  o_values;
  	Manager* mgr = Manager::Get();
  	g_criticalSection.lock();
! 	_return.retval =  mgr->SceneGetValues((::uint8 const) _sceneId, &o_values);
  	g_criticalSection.unlock();
+     for (int i=0; i< _return.retval; i++) _return.o_value.push_back(RemoteValueID(o_values[i]));
    }
  
    void SceneGetValueAsBool(Bool_Bool& _return, const int8_t _sceneId, const RemoteValueID& _valueId) {
  	Manager* mgr = Manager::Get();
  	g_criticalSection.lock();
***************
*** 1033,1044 ****
  	 mgr->GetDriverStatistics((::uint32 const) _homeId, (OpenZWave::Driver::DriverData*) &_return._data);
  	g_criticalSection.unlock();
    }
  
    void SendAllValues() {
!     // Your implementation goes here
!     printf("SendAllValues\n");
    }
  
  };
  
  // int main(int argc, char **argv) {
--- 1042,1054 ----
  	 mgr->GetDriverStatistics((::uint32 const) _homeId, (OpenZWave::Driver::DriverData*) &_return._data);
  	g_criticalSection.unlock();
    }
  
    void SendAllValues() {
!     g_criticalSection.lock();
!     send_all_values();
!     g_criticalSection.unlock();
    }
  
  };
  
  // int main(int argc, char **argv) {
