*** gen-cpp/RemoteManager_server.cpp.orig	2012-02-28 18:11:22.650899098 +0200
--- gen-cpp/RemoteManager_server.cpp.patched	2012-02-29 16:58:31.394945828 +0200
***************
*** 1,7 ****
  // Automatically generated OpenZWave::Manager_server wrapper
! // (c) 2011 Elias Karakoulakis <elias.karakoulakis@gmail.com>
  
  #include "RemoteManager.h"
  #include <protocol/TBinaryProtocol.h>
  #include <server/TSimpleServer.h>
  #include <transport/TServerSocket.h>
--- 1,7 ----
  // Automatically generated OpenZWave::Manager_server wrapper
! // (c) 2011-2012 Elias Karakoulakis <elias.karakoulakis@gmail.com>
  
  #include "RemoteManager.h"
  #include <protocol/TBinaryProtocol.h>
  #include <server/TSimpleServer.h>
  #include <transport/TServerSocket.h>
***************
*** 15,25 ****
  using boost::shared_ptr;
  
  using namespace  ::OpenZWave;
  
  void BeginControllerCommand_callback(OpenZWave::Driver::ControllerState  arg1, void*  arg2) {
! 	// FIXME: fill in the blanks (sorry!)
  }
  
  class RemoteManagerHandler : virtual public RemoteManagerIf {
   public:
    RemoteManagerHandler() {
--- 15,32 ----
  using boost::shared_ptr;
  
  using namespace  ::OpenZWave;
  
  void BeginControllerCommand_callback(OpenZWave::Driver::ControllerState  arg1, void*  arg2) {
!     g_criticalSection.lock();
!     //
!     STOMP::hdrmap headers;
!     headers["ControllerState"] = to_string<uint16_t>(arg1, std::hex);
!     string empty = ""  ;
!     stomp_client->send(*notifications_topic, headers, empty);
!     //
!     g_criticalSection.unlock();
  }
  
  class RemoteManagerHandler : virtual public RemoteManagerIf {
   public:
    RemoteManagerHandler() {
***************
*** 258,271 ****
  	_return =  mgr->GetNodeType((::uint32 const) _homeId, (::uint8 const) _nodeId);
  	g_criticalSection.unlock();
    }
  
    void GetNodeNeighbors(UInt32_ListByte& _return, const int32_t _homeId, const int8_t _nodeId) {
  	Manager* mgr = Manager::Get();
  	g_criticalSection.lock();
! 	_return.retval =  mgr->GetNodeNeighbors((::uint32 const) _homeId, (::uint8 const) _nodeId, (::uint8**) &_return._nodeNeighbors);
  	g_criticalSection.unlock();
    }
  
    void GetNodeManufacturerName(std::string& _return, const int32_t _homeId, const int8_t _nodeId) {
  	Manager* mgr = Manager::Get();
  	g_criticalSection.lock();
--- 265,283 ----
  	_return =  mgr->GetNodeType((::uint32 const) _homeId, (::uint8 const) _nodeId);
  	g_criticalSection.unlock();
    }
  
    void GetNodeNeighbors(UInt32_ListByte& _return, const int32_t _homeId, const int8_t _nodeId) {
+     uint8* arr;
  	Manager* mgr = Manager::Get();
  	g_criticalSection.lock();
! 	_return.retval =  mgr->GetNodeNeighbors((::uint32 const) _homeId, (::uint8 const) _nodeId, (::uint8**) &arr);
  	g_criticalSection.unlock();
+     if (_return.retval > 0) {
+         for (int i=0; i<_return.retval; i++) _return._nodeNeighbors.push_back(arr[i]);
+         delete arr;
+     } 
    }
  
    void GetNodeManufacturerName(std::string& _return, const int32_t _homeId, const int8_t _nodeId) {
  	Manager* mgr = Manager::Get();
  	g_criticalSection.lock();
***************
*** 523,537 ****
  	g_criticalSection.lock();
  	_return.retval =  mgr->GetValueListSelection(_id.toValueID(), (::int32*) &_return.o_value);
  	g_criticalSection.unlock();
    }
  
!   void GetValueListItems(Bool_ListString& _return, const RemoteValueID& _id) {
  	Manager* mgr = Manager::Get();
  	g_criticalSection.lock();
! 	_return.retval =  mgr->GetValueListItems(_id.toValueID(), (std::vector<std::string, std::allocator<std::string> >*) &_return.o_value);
  	g_criticalSection.unlock();
    }
  
    void GetValueFloatPrecision(Bool_UInt8& _return, const RemoteValueID& _id) {
  	Manager* mgr = Manager::Get();
  	g_criticalSection.lock();
--- 535,553 ----
  	g_criticalSection.lock();
  	_return.retval =  mgr->GetValueListSelection(_id.toValueID(), (::int32*) &_return.o_value);
  	g_criticalSection.unlock();
    }
  
!   void GetValueListItems(Bool_ListString& _return, const RemoteValueID& _id) {      
!     std::vector<std::string> o_values;
  	Manager* mgr = Manager::Get();
  	g_criticalSection.lock();
! 	_return.retval =  mgr->GetValueListItems(_id.toValueID(), &o_values);
  	g_criticalSection.unlock();
+     if (_return.retval > 0) {
+         for (int i=0; i< _return.retval; i++) _return.o_value.push_back(std::string(o_values[i]));
+     }
    }
  
    void GetValueFloatPrecision(Bool_UInt8& _return, const RemoteValueID& _id) {
  	Manager* mgr = Manager::Get();
  	g_criticalSection.lock();
***************
*** 700,713 ****
  	g_criticalSection.unlock();
  	return(function_result);
    }
  
    void GetAssociations(GetAssociationsReturnStruct& _return, const int32_t _homeId, const int8_t _nodeId, const int8_t _groupIdx) {
  	Manager* mgr = Manager::Get();
  	g_criticalSection.lock();
! 	_return.retval =  mgr->GetAssociations((::uint32 const) _homeId, (::uint8 const) _nodeId, (::uint8 const) _groupIdx, (::uint8**) &_return.o_associations);
  	g_criticalSection.unlock();
    }
  
    int8_t GetMaxAssociations(const int32_t _homeId, const int8_t _nodeId, const int8_t _groupIdx) {
  	Manager* mgr = Manager::Get();
  	g_criticalSection.lock();
--- 716,734 ----
  	g_criticalSection.unlock();
  	return(function_result);
    }
  
    void GetAssociations(GetAssociationsReturnStruct& _return, const int32_t _homeId, const int8_t _nodeId, const int8_t _groupIdx) {
+ 	uint8* o_associations;
  	Manager* mgr = Manager::Get();
  	g_criticalSection.lock();
! 	_return.retval =  mgr->GetAssociations((::uint32 const) _homeId, (::uint8 const) _nodeId, (::uint8 const) _groupIdx, (::uint8**) &o_associations);
  	g_criticalSection.unlock();
+     if (_return.retval > 0) {
+         for (int i=0; i<_return.retval; i++) _return.o_associations.push_back(o_associations[i]);
+         delete o_associations;
+     }
    }
  
    int8_t GetMaxAssociations(const int32_t _homeId, const int8_t _nodeId, const int8_t _groupIdx) {
  	Manager* mgr = Manager::Get();
  	g_criticalSection.lock();
***************
*** 774,787 ****
  	g_criticalSection.unlock();
  	return(function_result);
    }
  
    void GetAllScenes(GetAllScenesReturnStruct& _return) {
  	Manager* mgr = Manager::Get();
  	g_criticalSection.lock();
! 	_return.retval =  mgr->GetAllScenes((::uint8**) &_return._sceneIds);
  	g_criticalSection.unlock();
    }
  
    int8_t CreateScene() {
  	Manager* mgr = Manager::Get();
  	g_criticalSection.lock();
--- 795,813 ----
  	g_criticalSection.unlock();
  	return(function_result);
    }
  
    void GetAllScenes(GetAllScenesReturnStruct& _return) {
+     uint8* _sceneIds;
  	Manager* mgr = Manager::Get();
  	g_criticalSection.lock();
! 	_return.retval =  mgr->GetAllScenes((::uint8**) &_sceneIds);
  	g_criticalSection.unlock();
+     if (_return.retval>0) {
+         for (int i=0; i<_return.retval; i++) _return._sceneIds.push_back(_sceneIds[i]);
+         delete(_sceneIds);
+     }  
    }
  
    int8_t CreateScene() {
  	Manager* mgr = Manager::Get();
  	g_criticalSection.lock();
***************
*** 869,882 ****
  	g_criticalSection.unlock();
  	return(function_result);
    }
  
    void SceneGetValues(SceneGetValuesReturnStruct& _return, const int8_t _sceneId) {
  	Manager* mgr = Manager::Get();
  	g_criticalSection.lock();
! 	_return.retval =  mgr->SceneGetValues((::uint8 const) _sceneId, _return.o_value.toValueID());
  	g_criticalSection.unlock();
    }
  
    void SceneGetValueAsBool(Bool_Bool& _return, const int8_t _sceneId, const RemoteValueID& _valueId) {
  	Manager* mgr = Manager::Get();
  	g_criticalSection.lock();
--- 895,910 ----
  	g_criticalSection.unlock();
  	return(function_result);
    }
  
    void SceneGetValues(SceneGetValuesReturnStruct& _return, const int8_t _sceneId) {
+     std::vector<OpenZWave::ValueID>  o_values;
  	Manager* mgr = Manager::Get();
  	g_criticalSection.lock();
! 	_return.retval =  mgr->SceneGetValues((::uint8 const) _sceneId, &o_values);
  	g_criticalSection.unlock();
+     for (int i=0; i< _return.retval; i++) _return.o_value.push_back(RemoteValueID(o_values[i]));
    }
  
    void SceneGetValueAsBool(Bool_Bool& _return, const int8_t _sceneId, const RemoteValueID& _valueId) {
  	Manager* mgr = Manager::Get();
  	g_criticalSection.lock();
***************
*** 1033,1044 ****
  	 mgr->GetDriverStatistics((::uint32 const) _homeId, (OpenZWave::Driver::DriverData*) &_return._data);
  	g_criticalSection.unlock();
    }
  
    void SendAllValues() {
!     // Your implementation goes here
!     printf("SendAllValues\n");
    }
  
  };
  
  // int main(int argc, char **argv) {
--- 1061,1073 ----
  	 mgr->GetDriverStatistics((::uint32 const) _homeId, (OpenZWave::Driver::DriverData*) &_return._data);
  	g_criticalSection.unlock();
    }
  
    void SendAllValues() {
!     g_criticalSection.lock();
!     send_all_values();
!     g_criticalSection.unlock();
    }
  
  };
  
  // int main(int argc, char **argv) {
